### ISSUE TITLE
Track Functions Applied to a Signal

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#3043]
#### Describe the functionality you would like to see.
I think that it would be useful to save the functions that are applied to a signal. This is good practice and when given a dataset it would be useful to see if the data has been modified in any way and how that modification occurred.

I think that the easiest way to do this would be to create a decorator which saves the function in the metadata.

```python

def save_method(func,):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        n = max(self.metadata.General.function_history)
        self.metadata.General.function_history[n+1] = func: {"args": args, "kwargs:kwargs}
        return func(*args, **kwargs)
    return wrapper
 ```

Then you could decorate the code as such:

```python
@save_method
def sum(self, axis=None, out=None, rechunk=True):
#....
```
or if you wanted to limit which args that are saved

```python
@save_method(save=["axis"])
def sum(self, axis=None, out=None, rechunk=True):
#....
```

Honestly, I might need to play around with the actual implementation of this but it seems like it should be fairly easy to accomplish.

There are two potential issues:

1 - Each function should probably set what *args/**kwargs they want saved so as to not  balloon the metadata 
2 - The save_method function should not be called from inside of another save_method function.  

Figuring out the second case might be slightly tricky.  Off the top of my head we could have a variable is_tracking which is just set off and on at the end of the function. 

```python

def save_method(func,):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
       if self.is_tracking is False:
           self.is_tracking =True
           n = max(self.metadata.General.function_history)
           self.metadata.General.function_history[n+1] = func: {"args": args, "kwargs:kwargs}
           value =  func(*args, **kwargs)
           self.is_tracking=False
           return value
        else:
            return func(*args, **kwargs)
    return wrapper
 ```

**Comment by ericpre** [comment]
Should it be a workflow/data management thing?

**Comment by CSSFrancis** [comment]
> Should it be a workflow/data management thing?

Are you saying this is outside of the scope of hyperspy? 

Possibly, this might be overkill and could introduce some bad habits.  Ideally the result is saved alongside the workflow but that isn't always the case.

In reality there are only a couple of methods which we might want to track, and kind of do to some extent through changing the title or adding to the metadata or logging. Things like sum, mean, rebin, remove peaks, affine transformations in pyxem. Which adjust the raw data of some signal. Maybe for a find peaks method tracking exactly what parameters give that list of peaks is important.

Another alternative would be to possibly capture the logger output within the metadata for some sample.

**Comment by hakonanes** [comment]
I agree that a streamlined way to keep track of *all* operations that produced a result is good practice.

That said, aren't we doing so by storing Python scripts or Jupyter notebooks alongside our results? In these files, *all* operations are tracked, and all parameters are listed (besides default values, but these are documented in docstrings).

Furthermore, HyperSpy and its extension packages allow the user to freely overwrite the `BaseSignal.data` array. Such operations will not be tracked in the proposed HyperSpy functionality, right? But they are tracked in a script or notebook. Is an incomplete history in the metadata better than no history in the metadata? Personally, I would prefer to obtain a result produced by someone with HyperSpy with an accompanying script/notebook rather than a possibly incomplete history of operations in the metadata.
